rye/src/cli/publish.rs

```diff
@@ -68,96 +76,101 @@ pub fn execute(cmd: Args) -> Result<(), Error> {
 
-        maybe_decrypt(&secret, cmd.yes)?
-    } else {
-        echo!("No access token found, generate one at: https://pypi.org/manage/account/token/");
-        let token = if !cmd.yes {
-            prompt_for_token()?
+    // Token came from command line
+    let mut should_encrypt = token.is_some();
+
+    // Fallback prompts
+    if credentials.password.is_none() && !cmd.yes {
+        if credentials.username.is_none() {
+            credentials.username = prompt_username()?;
+        }
+
+        if is_unknown_repository(&repository) || is_default_repository(&repository) {
+            echo!("No access token found, generate one at: https://pypi.org/manage/account/token/");
+            credentials.password = prompt_token()?;
         } else {
-            "".to_string()
-        };
-        if token.is_empty() {
-            bail!("an access token is required")
+            credentials.password = prompt_password()?;
         }
-        let secret = Secret::new(token);
-        let maybe_encrypted = maybe_encrypt(&secret, cmd.yes)?;
-        let maybe_encoded = maybe_encode(&secret, &maybe_encrypted);
-        credentials[repository]["token"] = Item::Value(maybe_encoded.expose_secret().into());
+        should_encrypt = credentials.password.is_some();
+    }
 
-        secret
+    let passphrase = if cmd.yes || (!should_decrypt && !should_encrypt) {
+        None
+    } else if should_decrypt {
+        prompt_decrypt_passphrase()?
+    } else {
+        prompt_encrypt_passphrase()?
     };
 
-    credentials[repository]["repository-url"] = Item::Value(repository_url.to_string().into());
-    credentials[repository]["username"] = Item::Value(username.clone().into());
-    write_credentials(&credentials)?;
+    if repository.url.is_none() && !cmd.yes {
+        repository.url = prompt_repository_url()?;
+    }
+
+    // keyring fallback
+    if credentials.username.is_none() && repository.url.is_some() {
+        credentials.username = Some(DEFAULT_USERNAME.to_string())
+    }
+
+    if credentials.password.is_none() && !is_keyring_ready(&credentials, &repository) {
+        bail!(
+            "no configuration was found for repository '{}'",
+            repository.name.unwrap_or_default()
+        );
+    }
+
+    if !cmd.skip_save_credentials && repository.name.is_some() {
+        save_rye_credentials(
+            &mut credentials_file,
+            &credentials,
+            &repository,
+            should_encrypt,
+            passphrase.as_ref(),
+        )?;
+    }
 
     let mut publish_cmd = Command::new(get_venv_python_bin(&venv));
+
+    // Build Twine command
     publish_cmd
         .arg("-mtwine")
         .arg("--no-color")
         .arg("upload")
-        .args(files)
-        .arg("--username")
-        .arg(username)
-        .arg("--password")
-        .arg(token.expose_secret())
-        .arg("--repository-url")
-        .arg(repository_url.to_string());
+        .arg("--non-interactive")
```

Comment:
`--non-interactive` is used to make sure we're not triggering Twine interactive prompts.

```
--non-interactive     Do not interactively prompt for username/password if the required credentials are missing. (Can also be set via TWINE_NON_INTERACTIVE
                        environment variable.)
```